File: Project/Part-2 GoLang/RANSAC.go
/**
*@author: Rakshita Mathur
*@student number: 300215340
*@date: 03/11/2023
*@Course: CSI 2120 - programming paradigms
*@description: This program implements the RANSAC algorithm to find the best plane that fits a set of points.
*              The program takes as input a file containing the points in the format x y z, one point per line.
*              The program outputs the plane coefficients A, B, C, D and the number of points that are on the plane.
*              The program also outputs the time taken to run the algorithm.
*              The program takes as input the following parameters:
*              -i: input file name (default: PointCloud1) // do not include the extension
*              -c: confidence level (default: 0.99)
*              -p: percentage of points that are on the plane (default: 20)
*              -e: maximum distance between a point and the plane (default: 0.2)
*              Example run:  go run RANSAC.go PointCloud2 .99 20 .01
* 			  The program will run the RANSAC algorithm on the points in PointCloud2.xyz with a confidence level of 0.99, a percentage of points on the plane of 20% and a maximum distance of 0.01.
*
 */
package main

import (
	"bufio"
	"fmt"
	"log"
	"math"
	"math/rand"
	"os"
	"strconv"
	"strings"
)

type Point3D struct {
	X float64
	Y float64
	Z float64
}
type Plane3D struct {
	A float64
	B float64
	C float64
	D float64
}
type Plane3DwSupport struct {
	Plane3D
	SupportSize int
}

// Implement a function to read the XYZ file and return a slice of Point3D.
func ReadXYZ(filename string) []Point3D {
	// Open the file
	file, err := os.Open(filename)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	// Read the data from the file
	scanner := bufio.NewScanner(file)
	var points []Point3D
	var skippedFirstLine bool
	for scanner.Scan() {
		line := scanner.Text()
		if !skippedFirstLine {
			skippedFirstLine = true
			continue
		}
		fields := strings.Fields(line)
		if len(fields) != 3 {
			continue
		}
		x, err := strconv.ParseFloat(fields[0], 64)
		if err != nil {
			continue
		}
		y, err := strconv.ParseFloat(fields[1], 64)
		if err != nil {
			continue
		}
		z, err := strconv.ParseFloat(fields[2], 64)
		if err != nil {
			continue
		}
		points = append(points, Point3D{x, y, z})
	}
	return points
}

// Implement a function to save a slice of Point3D into an XYZ file.
func SaveXYZ(filename string, points []Point3D) {
	// Open the file for writing
	file, err := os.Create(filename)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	// Write each point to the file in XYZ format
	for _, point := range points {
		_, err = fmt.Fprintf(file, "%.6f %.6f %.6f\n", point.X, point.Y, point.Z)
		if err != nil {
			log.Fatal(err)
		}
	}
}

// Implement a function to compute the distance between two points in 3D space.
func (p1 *Point3D) GetDistance(p2 *Point3D) float64 {
	// compute the distance
	return math.Sqrt(math.Pow(p1.X-p2.X, 2) + math.Pow(p1.Y-p2.Y, 2) + math.Pow(p1.Z-p2.Z, 2))
}

// Implement a function to compute the plane defined by a set of three points.
// The plane is defined by the equation Ax + By + Cz + D = 0.
func GetPlane(points []Point3D) Plane3D {
	if len(points) != 3 {
		return Plane3D{}
	}
	// Compute the normal vector of the plane
	v1 := Point3D{X: points[1].X - points[0].X, Y: points[1].Y - points[0].Y, Z: points[1].Z - points[0].Z}
	v2 := Point3D{X: points[2].X - points[0].X, Y: points[2].Y - points[0].Y, Z: points[2].Z - points[0].Z}
	n := Point3D{X: v1.Y*v2.Z - v1.Z*v2.Y, Y: v1.Z*v2.X - v1.X*v2.Z, Z: v1.X*v2.Y - v1.Y*v2.X}
	norm := math.Sqrt(n.X*n.X + n.Y*n.Y + n.Z*n.Z)
	if norm == 0 {
		return Plane3D{}
	}
	n.X /= norm
	n.Y /= norm
	n.Z /= norm

	// Compute the plane coefficients
	plane := Plane3D{A: n.X, B: n.Y, C: n.Z}
	plane.D = -1 * (plane.A*points[0].X + plane.B*points[0].Y + plane.C*points[0].Z)

	return plane
}

// Implement a function to compute the number of required RANSAC iterations.
func GetNumberOfIterations(confidence float64, percentageOfPointsOnPlane float64) float64 {
	percentageOfPointsOnPlane = percentageOfPointsOnPlane / 100.0
	numberOfIterations := math.Ceil((math.Log(1-confidence) / math.Log(2)) / (math.Log(1-math.Pow(percentageOfPointsOnPlane, 3)) / math.Log(2)))
	return numberOfIterations
}

// Implement a function to compute the support of a plane in a set of points
func GetSupport(plane Plane3D, points []Point3D, eps float64) Plane3DwSupport {
	var supportSize int
	for _, point := range points {
		dist := point.GetDistance(&Point3D{X: -plane.D / plane.A, Y: -plane.D / plane.B, Z: -plane.D / plane.C})
		if dist <= eps {
			supportSize++
		}
	}
	return Plane3DwSupport{Plane3D: plane, SupportSize: supportSize}
}

// Implement a function to extract the points that support a given plane.
// Implement a function to extract the points that support a given plane.

func GetSupportingPoints(plane Plane3D, points []Point3D, eps float64) []Point3D {
	var supportingPoints []Point3D
	for _, point := range points {
		dist := math.Abs(plane.A*point.X+plane.B*point.Y+plane.C*point.Z+plane.D) / math.Sqrt(plane.A*plane.A+plane.B*plane.B+plane.C*plane.C)
		if dist <= eps {
			supportingPoints = append(supportingPoints, point)
		}
	}
	return supportingPoints
}

// Implement a function to create a new slice of points in which all points belonging to the plane have been removed.
func RemovePoints(points []Point3D, plane Plane3D, threshold float64) []Point3D {
	var newPoints []Point3D
	for _, point := range points {
		if math.Abs(plane.A*point.X+plane.B*point.Y+plane.C*point.Z+plane.D) > threshold {
			newPoints = append(newPoints, point)
		}
	}
	return newPoints
}

// Implement the pipeline, consisting of the following components:
// 1.Random point generator: randomly selects a point from the provided slice of Point3D.
func RandomPointGenerator(points []Point3D) <-chan Point3D {
	out := make(chan Point3D)
	go func() {
		for {
			out <- points[rand.Intn(len(points))]
		}
	}()
	return out
}

// 2.Triplet of points generator: reads Point3D instances and accumulates 3 points, then sends them as an array of Point3D.

func TripletOfPointsGenerator(in <-chan Point3D) <-chan []Point3D {
	out := make(chan []Point3D)
	go func() {
		var points []Point3D
		for {
			points = append(points, <-in)
			if len(points) == 3 {
				out <- points
				points = nil
			}
		}
	}()
	return out
}

// 3.TakeN: reads arrays of Point3D and resends them. It stops the pipeline after having received N arrays. Where N is the number of iterations float64.

func TakeN(in <-chan []Point3D, numberOfIterations float64) <-chan []Point3D {
	out := make(chan []Point3D)
	go func() {
		for i := 0; i < int(numberOfIterations); i++ {
			out <- <-in
		}
	}()
	return out
}

// 4.Plane estimator: reads arrays of three Point3D and computes the plane defined by these points. Using the function GetPlane.

func PlaneEstimator(in <-chan []Point3D) <-chan Plane3D {
	out := make(chan Plane3D)
	go func() {
		for {
			points := <-in
			out <- GetPlane(points)
		}
	}()
	return out
}

//5. Supporting point finder: counts the number of points in the provided slice of Point3D that supports the received 3D plane.

func SupportingPointFinder(in <-chan Plane3D, points []Point3D, threshold float64) <-chan Plane3DwSupport {
	out := make(chan Plane3DwSupport)
	go func() {
		for {
			plane := <-in
			support := GetSupport(plane, points, threshold)
			// out <- Plane3DwSupport{Plane3D, int}
			out <- Plane3DwSupport{plane, support.SupportSize}
		}
	}()
	return out
}

// 6.Fan In: multiplexes the results received from multiple channels into one output channel.

func FanIn(ins ...<-chan Plane3DwSupport) <-chan Plane3DwSupport {
	out := make(chan Plane3DwSupport)
	for _, in := range ins {
		go func(in <-chan Plane3DwSupport) {
			for {
				out <- <-in
			}
		}(in)
	}
	return out
}

// 7.Dominant plane identifier: receives Plane3DwSupport instances and keeps in memory the plane with the best support received so far.

func DominantPlaneIdentifier(in <-chan Plane3DwSupport) <-chan Plane3DwSupport {
	out := make(chan Plane3DwSupport)
	go func() {
		var best Plane3DwSupport
		for {
			plane := <-in
			if plane.SupportSize > best.SupportSize {
				best = plane
			}
			out <- best
		}
	}()
	return out
}

//In the main function, read the XYZ file, find the number of iterations required based on the specified confidence and percentage, create and start the RANSAC find dominant plane pipeline, save the supporting points of the identified dominant plane in a file named by appending _p to the input filename, and save the original point cloud without the supporting points of the dominant plane in a file named by appending _p0 to the input filename.

func main() {

	// read the input arguments
	if len(os.Args) != 5 {
		log.Fatal("Usage: RANSAC <input file> <confidence> <percentage> <eps>")
	}
	inputFilename := os.Args[1]
	confidence, err := strconv.ParseFloat(os.Args[2], 64)
	if err != nil {
		log.Fatal(err)
	}
	percentage, err := strconv.ParseFloat(os.Args[3], 64)
	if err != nil {
		log.Fatal(err)
	}
	eps, err := strconv.ParseFloat(os.Args[3], 64)
	if err != nil {
		log.Fatal(err)
	}

	// read the input file
	points := ReadXYZ(inputFilename)

	// find the number of iterations required based on the specified confidence and percentage
	iterations := GetNumberOfIterations(confidence, percentage)

	randomPointGenerator := RandomPointGenerator(points)

	tripletOfPointsGenerator := TripletOfPointsGenerator(randomPointGenerator)

	takeN := TakeN(tripletOfPointsGenerator, iterations)

	planeEstimator := PlaneEstimator(takeN)

	supportingPointFinder := SupportingPointFinder(planeEstimator, points, eps)

	fanIn := FanIn(supportingPointFinder)

	dominantPlaneIdentifier := DominantPlaneIdentifier(fanIn)

	plane := <-dominantPlaneIdentifier

	supportingPoints := GetSupport(plane.Plane3D, points, eps)
	// save the inline points of the identified dominant plane in a file named by appending _p to the input filename
	SaveXYZ(inputFilename+"_p", GetSupportingPoints(supportingPoints.Plane3D, points, eps))
	// save the original point cloud without the supporting points of the dominant plane in a file named by appending _p0 to the input filename
	SaveXYZ(inputFilename+"_p0", RemovePoints(points, plane.Plane3D, eps))
}

File: Project/Part-4 Prolog/ransac.pl
% File: ransac.pl
% Author: Rakshita Mathur
% Student ID: 300215340
% Date: 2023-04-10
% Description: Predicates for implementing the RANSAC (RANdom SAmple Consensus) algorithm in Prolog.



% Predicate to randomly select 3 points from a list of points
% random3points(+Points, -Point3)
% This predicate is true if Point3 is a triplet of points randomly selected from the list of points Points.
% The triplet of points is of the form [[x1,y1,z1], [x2,y2,z2], [x3,y3,z3]].
random3points(Points, Point3) :-
    length(Points, N),
    N >= 3, % Ensure that the list has at least 3 points
    random_permutation(Points, RandomPoints), % Randomly permute the list of points
    take(RandomPoints, 3, Point3). % Take the first 3 points from the permuted list

% Helper predicate to take first N elements from a list
take(_, 0, []).
take([X|Xs], N, [X|Ys]) :-
    N > 0,
    N1 is N - 1,
    take(Xs, N1, Ys).

% Predicate to compute the coefficients of a plane equation given 3 points
% plane(+Point3, -Plane)
% This predicate is true if Plane is the equation of the plane defined by the three points of the list Point3.
% The plane is specified by the list [a, b, c, d] from the equation ax + by + cz = d.
% The list of points is of the form [[x1,y1,z1], [x2,y2,z2], [x3,y3,z3]].
plane(Point3, Plane) :-
    length(Point3, 3), % Ensure that Point3 contains exactly 3 points
    Point3 = [[X1, Y1, Z1], [X2, Y2, Z2], [X3, Y3, Z3]], % Extract the coordinates of the three points
    % Compute the coefficients of the plane equation
    A is (Y2 - Y1) * (Z3 - Z1) - (Z2 - Z1) * (Y3 - Y1),
    B is (Z2 - Z1) * (X3 - X1) - (X2 - X1) * (Z3 - Z1),
    C is (X2 - X1) * (Y3 - Y1) - (Y2 - Y1) * (X3 - X1),
    D is -(A * X1 + B * Y1 + C * Z1),
    Plane = [A, B, C, D]. % Construct the plane equation as a list [a, b, c, d]

% Predicate to find the support points of a plane given a list of points and an epsilon value
% support(+Plane, +Points, +Eps, -N)
% This predicate is true if the support of plane Plane is composed of N points from the list of points Point3
% when the distance Eps is used.
support(Plane, Points, Eps, N) :-
    length(Points, NumPoints), % Get the total number of points in the list
    NumPoints >= N, % Ensure that there are enough points to form a support of size N
    find_support_points(Plane, Points, Eps, SupportPoints), % Find the support points
    length(SupportPoints, N). % Check if the number of support points is N


% Helper predicate to find the support points for a given plane
find_support_points(_, [], _, []).
find_support_points(Plane, [Point|Points], Eps, SupportPoints) :-
    % Extract the coordinates of the current point
    Point = [X, Y, Z],
    % Compute the distance from the current point to the plane
    Dist is abs(Plane[0] * X + Plane[1] * Y + Plane[2] * Z + Plane[3]) / sqrt(Plane[0] * Plane[0] + Plane[1] * Plane[1] + Plane[2] * Plane[2]),
    % Check if the distance is within the given epsilon value
    Dist =< Eps,
    % Recursively find the support points in the remaining list of points
    find_support_points(Plane, Points, Eps, RemainingSupportPoints),
    % Construct the list of support points
    SupportPoints = [Point|RemainingSupportPoints].

% Predicate to calculate the number of iterations required by RANSAC with given confidence and percentage parameters
% ransac_number_of_iterations(+Confidence, +Percentage, -N)
% This predicate is true if N is the number of iterations required by RANSAC with parameters Confidence and Percentage
ransac_number_of_iterations(Confidence, Percentage, N) :-
    Confidence > 0, % Ensure that Confidence is a positive value
    Confidence < 1, % Ensure that Confidence is less than 1
    Percentage > 0, % Ensure that Percentage is a positive value
    Percentage < 1, % Ensure that Percentage is less than 1
    NumPoints is round(log(1 - Confidence) / log(1 - Percentage^3)), % Compute the number of iterations using the formula
    N is max(1, NumPoints). % Set N to be at least 1, as per the problem description

%test case 

% Test cases for random3points/2 predicate
test(random3points, 1) :-
    Points = [[0,0,0], [-5.1323336,-4.089636333,0.243960825], [-5.141415625,-4.020067234,0.242623445]],
    random3points(Points, Point3),
    length(Point3, 3),
    writeln('Randomly selected 3 points:'),
    writeln(Point3).

test(random3points, 2) :-
    Points = [[1,1,1], [2,2,2], [3,3,3], [4,4,4]],
    random3points(Points, Point3),
    length(Point3, 3),
    writeln('Randomly selected 3 points:'),
    writeln(Point3).

% Test cases for plane/2 predicate
test(plane, 1) :-
    Point3 = [[0,0,0], [1,1,1], [2,2,2]],
    plane(Point3, Plane),
    writeln('Plane equation:'),
    writeln(Plane).

test(plane, 2) :-
    Point3 = [[-1,-1,-1], [0,0,0], [1,1,1]],
    plane(Point3, Plane),
    writeln('Plane equation:'),
    writeln(Plane).

% Test cases for support/4 predicate
test(support, 1) :-
    Plane = [1,1,1,0],
    Points = [[0,0,0], [1,1,1], [2,2,2], [3,3,3]],
    Eps = 0.1,
    N = 3,
    support(Plane, Points, Eps, N),
    writeln('Support points for the plane:'),
    writeln(Points).

test(support, 2) :-
    Plane = [0,0,1,5],
    Points = [[0,0,5], [1,1,6], [2,2,7], [3,3,8]],
    Eps = 0.2,
    N = 2,
    support(Plane, Points, Eps, N),
    writeln('Support points for the plane:'),
    writeln(Points).

% Test cases for ransac_number_of_iterations/3 predicate
test(ransac_number_of_iterations, 1) :-
    Confidence = 0.99,
    Percentage = 0.3,
    ransac_number_of_iterations(Confidence, Percentage, N),
    writeln('Number of iterations for RANSAC:'),
    writeln(N).

test(ransac_number_of_iterations, 2) :-
    Confidence = 0.95,
    Percentage = 0.2,
    ransac_number_of_iterations(Confidence, Percentage, N),
    writeln('Number of iterations for RANSAC:'),
    writeln(N).

% Instruction to run the test cases
% ?- test(random3points, N).
% ?- test(plane, N).
% ?- test(support, N).
% ?- test(ransac_number_of_iterations, N).

